### ISSUE TITLE
Kiali wizards not working on Istio 1.4-beta

### ISSUE THREAD

**Issue by israel-hdez** [istio/istio#18558]
**Bug description**

There is an issue in Kiali when using configuration Wizards available at the Service Details page. I'm talking about this:

!
**Expected behavior**
Kiali wizards should create the VirtualServices, DestinationRules and any other required objects to properly apply que requested config in the Wizards.

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
kubernets 1.15.2
Istio 1.4.0-beta-1
Tried on both Kiali v1.4.2 (currently shipped with Istio1.4) and also upgrading to latest `master` Kiali code.

**How was Istio installed?**
istioctl manifest apply --set profile=demo

**Environment where bug was observed (cloud vendor, OS, etc)**
minikube

**Comment by israel-hdez** [comment]
The istio-policy-bot removed my image :(
This is the image:
![](https://user-images.githubusercontent.com/23639005/68043517-17be7900-fc9b-11e9-8e7a-5499adcd9d48.png)

**Comment by israel-hdez** [comment]
The istio-policy-bot truncated my description. Please check the history to see the original description.

**Comment by howardjohn** [comment]
Issue on helm install, running k8s 1.15. Suspected: #16320 @Jason-ZW

**Comment by howardjohn** [comment]
Error is:

```
VirtualService.networking.istio.io "alpine" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"networking.istio.io/v1alpha3", "kind":"VirtualService", "metadata":map[string]interface {}{"creationTimestamp":"2019-11-01T20:14:53Z", "generation":1, "labels":map[string]interface {}{"kiali_wizard":"weighted_routing"}, "name":"alpine", "namespace":"default", "uid":"8c754124-c631-4c1b-98de-256770c8ecc3"}, "spec":map[string]interface {}{"gateways":interface {}(nil), "hosts":[]interface {}{"alpine"}, "http":[]interface {}{map[string]interface {}{"route":[]interface {}{}}}}}: validation failure list:
spec.gateways in body must be of type array: "null"
```

**Comment by howardjohn** [comment]
This doesn't look like a valid VirtualService... I think this is equivalent to
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: reviews-route
spec:
  gateways: ~
  hosts: [alpine]
  http:
  - route: [{}]
```

**Comment by israel-hdez** [comment]
I just tried with this YAML:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: reviews-route
spec:
  hosts:
  - reviews
  gateways: null
  http:
    - route:
      - destination:
          host: reviews
```

Note the `gateways: null` line.
This YAML gets rejected in Istio1.4 but is accepted in Istio1.3. By removing that attribute from the YAML, the VS is accepted in Istio1.4.

A similar issue happens with DestinationRule CRD.

**Comment by howardjohn** [comment]
```
Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"DestinationRule.networking.istio.io \"alpine\" is invalid: []: Invalid value: map[string]interface {}{\"apiVersion\":\"networking.istio.io/v1alpha3\", \"kind\":\"DestinationRule\", \"metadata\":map[string]interface {}{\"creationTimestamp\":\"2019-11-01T20:24:41Z\", \"generation\":1, \"labels\":map[string]interface {}{\"kiali_wizard\":\"weighted_routing\"}, \"name\":\"alpine\", \"namespace\":\"default\", \"uid\":\"2a160a32-3079-467a-a97f-3a70bdc3752d\"}, \"spec\":map[string]interface {}{\"host\":\"alpine\", \"subsets\":[]interface {}{}, \"trafficPolicy\":interface {}(nil)}}: validation failure list:\nspec.trafficPolicy in body must be of type object: \"null\"","reason":"Invalid","details":{"name":"alpine","group":"networking.istio.io","kind":"DestinationRule","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: map[string]interface {}{\"apiVersion\":\"networking.istio.io/v1alpha3\", \"kind\":\"DestinationRule\", \"metadata\":map[string]interface {}{\"creationTimestamp\":\"2019-11-01T20:24:41Z\", \"generation\":1, \"labels\":map[string]interface {}{\"kiali_wizard\":\"weighted_routing\"}, \"name\":\"alpine\", \"namespace\":\"default\", \"uid\":\"2a160a32-3079-467a-a97f-3a70bdc3752d\"}, \"spec\":map[string]interface {}{\"host\":\"alpine\", \"subsets\":[]interface {}{}, \"trafficPolicy\":interface {}(nil)}}: validation failure list:\nspec.trafficPolicy in body must be of type object: \"null\"","field":"[]"}]},"code":422}
```

DR is rejected as well

**Comment by howardjohn** [comment]
```yaml
cat <<EOF | kaf -
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: reviews-route
spec:
  hosts:
  - reviews
  gateways: nullcat <<EOF | kaf -
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: reviews-route
spec:
  hosts:
  - reviews
  gateways: null
  http:
    - route:
      - destination:
          host: reviews
EOF
  http:
    - route:
      - destination:
          host: reviews
EOF
```

works for me... which is weird because its being rejected from kiali but not kubectl somehow?

kube version: `Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-20T18:57:36Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}` (running kind)

**Comment by howardjohn** [comment]
Oh interesting, if I modify a VS (reviews-route) it is allowed, but if I make a new one its rejected

**Comment by geeknoid** [comment]
PS: Seems the policybot did some slash and burn on this issue's initial description. You can recover the original text via the pull down menu near the description area. I'm investigating what happened there, but I bet it has something to do with the embedded picture.

**Comment by jasonwzm** [comment]
> I just tried with this YAML:
> 
> ```yaml
> apiVersion: networking.istio.io/v1alpha3
> kind: VirtualService
> metadata:
>   name: reviews-route
> spec:
>   hosts:
>   - reviews
>   gateways: null
>   http:
>     - route:
>       - destination:
>           host: reviews
> ```
> 
> Note the `gateways: null` line.
> This YAML gets rejected in Istio1.4 but is accepted in Istio1.3. By removing that attribute from the YAML, the VS is accepted in Istio1.4.
> 
> A similar issue happens with DestinationRule CRD.

`gateways: null` should be `gateways: []`. The error shows: `spec.gateways in body must be of type array: "null"`

**Comment by jasonwzm** [comment]
@israel-hdez Please also let me know if there are other schema validation errors you are seeing from your side.

**Comment by israel-hdez** [comment]
Is there any chance that this issue can block Istio 1.4 release?
We are working on a fix in Kiali and want to do a release as quickly as possible. So, I'm asking just in case.

**Comment by jmazzitelli** [comment]
Kiali team is feverishly working on a fix for this -- see https://github.com/kiali/kiali/pull/1870 -- in the hopes of getting this into Istio 1.4 release. If we can do it in time, we'd need to update the Istio installer stuff so it pulls in this new Kiali release (right now it is pulling in Kiali v1.4.2).

cc @jwendell

**Comment by howardjohn** [comment]
@jmazzitelli are there major changes since Kiali 1.4.2 and kiali latest? It may be best to start doing testing of kiali latest with Istio 1.4 so that we can ensure that upgrade is fine, then pick up the CRD fix

**Comment by jmazzitelli** [comment]
@howardjohn there are some major changes since v1.4.2 yes.

**Comment by jmazzitelli** [comment]
Once we get a new 1.4.beta2, I'll be testing to ensure the installer works (we need the CNI fixes). We'll also be doing our own testing over here.

**Comment by jmazzitelli** [comment]
(BTW: folks can test the latest 1.8 release which went out friday... we'll be building this fix into a new 1.9 that we will release for this fix. We usually release on a 3-week cadence but for this fix, we can release 1.9 this week, if we can get it into istio 1.4. If not, we'll probably just hold off and keep with the regular release cadence.

**Comment by jmazzitelli** [comment]
This is also related: https://github.com/kiali/kiali/pull/1888

**Comment by jmazzitelli** [comment]
When we upgrade to latest Kiali with this [PR](https://github.com/istio/installer/pull/503), this should be fixed.

cc @lucasponce @israel-hdez so they can confirm this.

**Comment by israel-hdez** [comment]
Kiali is fixed and Istio 1.4 is updated to install the fixed Kiali.
Thus, I'm closing this issue.
